### ISSUE TITLE
Flat arrays in internal representation

### ISSUE THREAD

**Issue by Athas** [HIPERFIT/futhark#101]
Currently, our internal representation supports nested (regular) arrays with extensive type information.  This is supported by frequent use of `reshape` operations, and the type checker is ready to punish any transgressions against strict shape correctness.  I am thinking about changing this.

First, a `reshape` is completely free at runtime, so that's not the problem.  The problem is that these semantically insignificant `reshape` operations obscure the computation taking place, and makes it harder to simplify and generate efficient code.

As an example, consider an operation done by the awkwardly named `KernelBabysitting` module (it's really responsible for frobbing array representations to ensure coalesced memory access in GPU kernels).  One important operation is taking an array of size `N`, treating it as a two-dimensional array of size `(N/K, K)`, transposing the two dimensions, and collapsing them back into one of size N.  The result is that this one-dimensional array is now transposed in memory, which is what we need for efficient traversal in the per-thread sequential parts of scan and reduce kernels.  (There's also some padding going on to ensure that N is divisible by K, but that's not the issue.)

In Futhark, this operation currently looks like this, assuming the original array is `a`:

```
let [[int,N/K],K] b = reshape((N/K, K), a)
let [[int,K],N/K] c = transpose(b)
let [int,N] d = reshape((N), c)
```

Currently this results in a rather gross index function when accessing elements of `d`.  We can add some intelligence (and I will) to detect that the two reshapes are respectively an explosion and implosion of a single dimension, but this is brittle.  Furthermore, the prevalence of `reshape` operations also prevent issues elsewhere, such as in fusion.

If we think a little about it, array shapes are really mostly used for two things:
- SOAC iteration spaces
- Explicit index operations

The former can be replaced by annotating every SOAC with the shape it interprets its input arrays as having.  The latter is not really necessary until code generation, at which point we have index functions, but can be annotated the same way as SOACs.  In short, I am suggesting moving shape annotations from where arrays are _created_ to where they are _used_.

The biggest problem is what to do about assertions, as currently we use `reshape` operations as the foundation of our `zip` safety.  We can continue to use our old assertion mechanism, but we will need somewhere else to attach certificates.  Possibly we can simply institute the rule that assert-expressions are never removed by dead code elimination (unless they are unreachable).  I don't clearly remember why we went away from that approach.

We will also still need a notion of nested arrays for existential return types from functions and function-ish things like branches and sequential loops.

If we think this is a good idea, we can move in this direction in steps by making individual operations parametrised by the shapes they accept and produce.  For example, the above could be written:

```
let [int,N] d = transpose((N/K, k), (N), a)
```

This saves us from doing a big refactoring in one single step.

Finally, all this only affects the internal language.  The external is unchanged.  Nested arrays are a useful human notation.

**Comment by Athas** [comment]
As an experiment, I modified part of our internal `reduceKernel` and `scanKernel` constructs to pretend that the outer dimension of their input arrays are in fact two dimensions, corresponding to the number of threads and the per-thread chunk size.  This permits much more natural index functions, without any reshape ugliness.  This is the old input fetch for a simple multiply-scan:

```
            y_29 = *(__global int *) &mem_111[((i_27 - i_27 /
                                                per_thread_elements_39 *
                                                per_thread_elements_39) *
                                               num_threads_36 + i_27 /
                                               per_thread_elements_39) * 4];
```

And this is what it looks like now:

```
            y_29 = *(__global int *) &mem_111[(elements_scanned_148 *
                                               num_threads_36 + global_id_139) *
                                              4];
```

This is much more pleasing to the eye, but I'm afraid that's the primary benefit in this case.  I measured a performance improvement of around 1-2% in most cases.  In retrospect, the reason is obvious: any global memory access is going to involve so much latency (likely 10-20 cycles) that the integer calculations are easily hidden.  I am trying to think of a contrived case where this would make a larger difference, but I think programs that are index-computation-bound but not memory-bound are a pretty rare breed.

Still, it is definitely worthwhile to have more readable output programs.
